From 7a6e7b028fa00950d36530ebf85208a4290a20ce Mon Sep 17 00:00:00 2001
From: Chetan Mistry <chetan.mistry@morsemicro.com>
Date: Mon, 5 May 2025 08:15:26 +1000
Subject: [PATCH] Configure NetworkManager to talk to 2 WPA Supplicants

This change allows NetworkManager to comminucate with 2
WPA Supplicant services (wpa_supplicant and wpa_supplicant_s1g).
The intended usage is that wpa_supplicant_s1g handles all wifi connections
on MorseMicro/S1G radios and wpa_supplicant handles everything else.
It works by extending the supplicant manager and interface manager structures
to contain the extra information needed to communicate with the additional
supplicant.
It also dynamically identifies which interfaces are running on morse radios.

Signed-off-by: Chetan Mistry <chetan.mistry@morsemicro.com>
---
 src/core/supplicant/nm-supplicant-interface.c | 199 +++++---
 src/core/supplicant/nm-supplicant-manager.c   | 424 ++++++++++++++++--
 src/core/supplicant/nm-supplicant-manager.h   |   3 +
 src/core/supplicant/nm-supplicant-types.h     |  19 +-
 4 files changed, 540 insertions(+), 105 deletions(-)

diff --git a/src/core/supplicant/nm-supplicant-interface.c b/src/core/supplicant/nm-supplicant-interface.c
index eb7a40a..0eb99f1 100644
--- a/src/core/supplicant/nm-supplicant-interface.c
+++ b/src/core/supplicant/nm-supplicant-interface.c
@@ -88,6 +88,7 @@ typedef struct _NMSupplicantInterfacePrivate {
 
     GDBusConnection *dbus_connection;
     NMRefString     *name_owner;
+    NMRefString     *name_owner_s1g;
     NMRefString     *object_path;
 
     char *ifname;
@@ -340,6 +341,19 @@ security_from_vardict(GVariant *security)
 
 /*****************************************************************************/
 
+/**
+ * Simple helper function to determine if the interface path is an S1G target
+ */
+bool is_path_s1g(const char* path){
+  const char* search_target = "s1g";
+    for(size_t i = 0; i < strlen(path) - strlen(search_target); i++){
+      if(strncmp(path+i,search_target, strlen(search_target)) == 0){
+        return true;
+      }
+    }
+    return false;
+}
+
 static void
 _dbus_connection_call(NMSupplicantInterface *self,
                       const char            *interface_name,
@@ -354,8 +368,11 @@ _dbus_connection_call(NMSupplicantInterface *self,
 {
     NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
+
     g_dbus_connection_call(priv->dbus_connection,
-                           priv->name_owner->str,
+                           bus_name,
                            priv->object_path->str,
                            interface_name,
                            method_name,
@@ -485,8 +502,9 @@ _remove_network(NMSupplicantInterface *self)
         return;
 
     net_path = g_steal_pointer(&priv->net_path);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple(self,
-                                 NM_WPAS_DBUS_IFACE_INTERFACE,
+                                 is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                  "RemoveNetwork",
                                  g_variant_new("(o)", net_path),
                                  G_VARIANT_TYPE("()"),
@@ -497,7 +515,7 @@ _remove_network(NMSupplicantInterface *self)
                                      DBUS_INTERFACE_PROPERTIES,
                                      "Set",
                                      g_variant_new("(ssv)",
-                                                   NM_WPAS_DBUS_IFACE_INTERFACE,
+                                                   is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                                    "ApIsolate",
                                                    g_variant_new_string("0")),
                                      G_VARIANT_TYPE("()"),
@@ -847,10 +865,13 @@ _bss_info_add(NMSupplicantInterface *self, const char *object_path)
     c_list_link_tail(&priv->bss_initializing_lst_head, &bss_info->_bss_lst);
     g_hash_table_add(priv->bss_idx, bss_info);
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
+
     nm_dbus_connection_call_get_all(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     bss_info->bss_path->str,
-                                    NM_WPAS_DBUS_IFACE_BSS,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_BSS_S1G : NM_WPAS_DBUS_IFACE_BSS,
                                     5000,
                                     bss_info->_init_cancellable,
                                     _bss_info_get_all_cb,
@@ -1045,10 +1066,12 @@ _peer_info_add(NMSupplicantInterface *self, const char *object_path)
     c_list_link_tail(&priv->peer_initializing_lst_head, &peer_info->_peer_lst);
     g_hash_table_add(priv->peer_idx, peer_info);
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     nm_dbus_connection_call_get_all(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     peer_info->peer_path->str,
-                                    NM_WPAS_DBUS_IFACE_PEER,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_PEER_S1G : NM_WPAS_DBUS_IFACE_PEER,
                                     5000,
                                     peer_info->_init_cancellable,
                                     _peer_info_get_all_cb,
@@ -1094,9 +1117,11 @@ set_state_down(NMSupplicantInterface *self,
     nm_assert(priv->state != NM_SUPPLICANT_INTERFACE_STATE_DOWN);
     nm_assert(!c_list_is_empty(&self->supp_lst));
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     _LOGD("remove interface \"%s\" on %s (%s)%s",
           priv->object_path->str,
-          priv->name_owner->str,
+          bus_name,
           reason,
           force_remove_from_supplicant ? " (remove in wpa_supplicant)" : "");
 
@@ -1110,7 +1135,7 @@ set_state_down(NMSupplicantInterface *self,
 
     if (force_remove_from_supplicant) {
         _nm_supplicant_manager_dbus_call_remove_interface(priv->supplicant_manager,
-                                                          priv->name_owner->str,
+                                                          bus_name,
                                                           priv->object_path->str);
     }
 
@@ -1353,12 +1378,14 @@ _starting_check_ready(NMSupplicantInterface *self)
 
     /* Other global properties are set in constructed() because they don't
      * depend on interface capabilities. */
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     if (_get_capability(priv, NM_SUPPL_CAP_TYPE_SAE) == NM_TERNARY_TRUE) {
         _LOGD("enabling SAE-H2E (SaePwe=2)");
         nm_dbus_connection_call_set(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     priv->object_path->str,
-                                    NM_WPAS_DBUS_IFACE_INTERFACE,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                     "SaePwe",
                                     g_variant_new_string("2"),
                                     DBUS_TIMEOUT_MSEC,
@@ -1481,10 +1508,12 @@ nm_supplicant_interface_set_bridge(NMSupplicantInterface *self, const char *brid
 
     _LOGT("set bridge %s%s%s", NM_PRINT_FMT_QUOTE_STRING(bridge));
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     nm_dbus_connection_call_set(priv->dbus_connection,
-                                priv->name_owner->str,
+                                bus_name,
                                 priv->object_path->str,
-                                NM_WPAS_DBUS_IFACE_INTERFACE,
+                                is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                 "BridgeIfname",
                                 g_variant_new_string(bridge ?: ""),
                                 DBUS_TIMEOUT_MSEC,
@@ -1593,20 +1622,22 @@ _p2p_group_set_path(NMSupplicantInterface *self, const char *path)
     nm_ref_string_unref(priv->p2p_group_path);
     priv->p2p_group_path = g_steal_pointer(&group_path);
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     if (priv->p2p_group_path) {
         priv->p2p_group_properties_cancellable = g_cancellable_new();
         priv->p2p_group_properties_changed_id =
             nm_dbus_connection_signal_subscribe_properties_changed(priv->dbus_connection,
-                                                                   priv->name_owner->str,
+                                                                   bus_name,
                                                                    priv->p2p_group_path->str,
-                                                                   NM_WPAS_DBUS_IFACE_GROUP,
+                                                                   is_s1g ? NM_WPAS_DBUS_IFACE_GROUP_S1G : NM_WPAS_DBUS_IFACE_GROUP,
                                                                    _p2p_group_properties_changed_cb,
                                                                    self,
                                                                    NULL);
         nm_dbus_connection_call_get_all(priv->dbus_connection,
-                                        priv->name_owner->str,
+                                        bus_name,
                                         priv->p2p_group_path->str,
-                                        NM_WPAS_DBUS_IFACE_GROUP,
+                                        is_s1g ? NM_WPAS_DBUS_IFACE_GROUP_S1G : NM_WPAS_DBUS_IFACE_GROUP,
                                         5000,
                                         priv->p2p_group_properties_cancellable,
                                         _p2p_group_properties_get_all_cb,
@@ -1700,9 +1731,11 @@ _wps_handle_set_pc_cb(GVariant *res, GError *error, gpointer user_data)
     else
         _LOGW("wps: ProcessCredentials failed to set (%s), starting...", error->message);
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     wps_data->signal_id = g_dbus_connection_signal_subscribe(priv->dbus_connection,
-                                                             priv->name_owner->str,
-                                                             NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
+                                                             bus_name,
+                                                             is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_WPS_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
                                                              "Credentials",
                                                              priv->object_path->str,
                                                              NULL,
@@ -1732,7 +1765,7 @@ _wps_handle_set_pc_cb(GVariant *res, GError *error, gpointer user_data)
         wps_data->cancellable = g_cancellable_new();
 
     _dbus_connection_call(self,
-                          NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
+                          bus_name,
                           "Start",
                           g_variant_new("(a{sv})", &start_args),
                           G_VARIANT_TYPE("(a{sv})"),
@@ -1751,10 +1784,12 @@ _wps_call_set_pc(NMSupplicantInterface *self, WpsData *wps_data)
     if (!wps_data->cancellable)
         wps_data->cancellable = g_cancellable_new();
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
     nm_dbus_connection_call_set(priv->dbus_connection,
-                                priv->name_owner->str,
+                                bus_name,
                                 priv->object_path->str,
-                                NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
+                                is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_WPS_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
                                 "ProcessCredentials",
                                 g_variant_new_boolean(TRUE),
                                 5000,
@@ -1860,8 +1895,11 @@ _wps_start(NMSupplicantInterface *self, const char *type, const char *bssid, con
         nm_clear_g_cancellable(&wps_data->cancellable);
         nm_clear_g_dbus_connection_signal(priv->dbus_connection, &wps_data->signal_id);
 
+    
+        bool is_s1g = is_path_s1g(priv->object_path->str);
+
         _dbus_connection_call(self,
-                              NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
+                              is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_WPS_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_WPS,
                               "Cancel",
                               NULL,
                               G_VARIANT_TYPE("()"),
@@ -2141,12 +2179,14 @@ nm_supplicant_interface_disconnect(NMSupplicantInterface *self)
     g_return_if_fail(NM_IS_SUPPLICANT_INTERFACE(self));
 
     priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    
+    bool is_s1g = is_path_s1g(priv->object_path->str);
 
     /* Disconnect from the current AP */
     if ((priv->state >= NM_SUPPLICANT_INTERFACE_STATE_SCANNING)
         && (priv->state <= NM_SUPPLICANT_INTERFACE_STATE_COMPLETED)) {
         _dbus_connection_call_simple(self,
-                                     NM_WPAS_DBUS_IFACE_INTERFACE,
+                                     is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                      "Disconnect",
                                      NULL,
                                      G_VARIANT_TYPE("()"),
@@ -2197,8 +2237,10 @@ nm_supplicant_interface_disconnect_async(NMSupplicantInterface            *self,
     g_return_if_fail(NM_IS_SUPPLICANT_INTERFACE(self));
     g_return_if_fail(callback);
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call(self,
-                          NM_WPAS_DBUS_IFACE_INTERFACE,
+                          is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                           "Disconnect",
                           NULL,
                           G_VARIANT_TYPE("()"),
@@ -2231,9 +2273,10 @@ static void
 assoc_call_select_network(NMSupplicantInterface *self)
 {
     NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
-
+    
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call(self,
-                          NM_WPAS_DBUS_IFACE_INTERFACE,
+                          is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                           "SelectNetwork",
                           g_variant_new("(o)", priv->net_path),
                           G_VARIANT_TYPE("()"),
@@ -2299,9 +2342,14 @@ assoc_add_network_cb(GObject *source, GAsyncResult *result, gpointer user_data)
 
     res = g_dbus_connection_call_finish(G_DBUS_CONNECTION(source), result, &error);
 
+
+    bool is_s1g = false;
+    is_s1g = is_path_s1g(object_path->str);
+
     if (!assoc_data) {
         if (!error) {
             const char *net_path;
+            const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
 
             /* the assoc-request was already cancelled, but the AddNetwork request succeeded.
              * Cleanup the created network.
@@ -2310,9 +2358,9 @@ assoc_add_network_cb(GObject *source, GAsyncResult *result, gpointer user_data)
              * and leaves the mainloop. During program shutdown, we may orphan networks. */
             g_variant_get(res, "(&o)", &net_path);
             g_dbus_connection_call(G_DBUS_CONNECTION(source),
-                                   name_owner->str,
+                                   bus_name,
                                    object_path->str,
-                                   NM_WPAS_DBUS_IFACE_INTERFACE,
+                                   is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                    "RemoveNetwork",
                                    g_variant_new("(o)", net_path),
                                    G_VARIANT_TYPE("()"),
@@ -2350,11 +2398,12 @@ assoc_add_network_cb(GObject *source, GAsyncResult *result, gpointer user_data)
         return;
     }
 
+
     g_hash_table_iter_init(&iter, blobs);
     while (g_hash_table_iter_next(&iter, (gpointer) &blob_name, (gpointer) &blob_data)) {
         _dbus_connection_call(
             self,
-            NM_WPAS_DBUS_IFACE_INTERFACE,
+            is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
             "AddBlob",
             g_variant_new("(s@ay)", blob_name, nm_g_bytes_to_variant_ay(blob_data)),
             G_VARIANT_TYPE("()"),
@@ -2391,9 +2440,10 @@ add_network(NMSupplicantInterface *self)
                                          "supplicant-add-network");
     priv->assoc_data->add_network_data = add_network_data;
 
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call(
         self,
-        NM_WPAS_DBUS_IFACE_INTERFACE,
+        is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
         "AddNetwork",
         g_variant_new("(@a{sv})", nm_supplicant_config_to_variant(priv->assoc_data->cfg)),
         G_VARIANT_TYPE("(o)"),
@@ -2554,14 +2604,16 @@ nm_supplicant_interface_assoc(NMSupplicantInterface       *self,
         assoc_data->fail_on_idle_id = g_idle_add(assoc_fail_on_idle_cb, self);
         return;
     }
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
 
     assoc_data->cancellable = g_cancellable_new();
     assoc_data->calls_left++;
     nm_dbus_connection_call_set(
         priv->dbus_connection,
-        priv->name_owner->str,
+        bus_name,
         priv->object_path->str,
-        NM_WPAS_DBUS_IFACE_INTERFACE,
+        is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
         "ApScan",
         g_variant_new_uint32(nm_supplicant_config_get_ap_scan(priv->assoc_data->cfg)),
         DBUS_TIMEOUT_MSEC,
@@ -2574,9 +2626,9 @@ nm_supplicant_interface_assoc(NMSupplicantInterface       *self,
      * reauthentication after only 8:24 hours (70% of the lifetime). */
     assoc_data->calls_left++;
     nm_dbus_connection_call_set(priv->dbus_connection,
-                                priv->name_owner->str,
+                                bus_name,
                                 priv->object_path->str,
-                                NM_WPAS_DBUS_IFACE_INTERFACE,
+                                is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                 "Dot11RSNAConfigPMKLifetime",
                                 g_variant_new_take_string(g_strdup_printf("%u", PMK_LIFETIME_SEC)),
                                 DBUS_TIMEOUT_MSEC,
@@ -2599,9 +2651,9 @@ nm_supplicant_interface_assoc(NMSupplicantInterface       *self,
          * need first a Get call. It seems simpler to just set the value
          * we want. */
         nm_dbus_connection_call_set(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     priv->object_path->str,
-                                    NM_WPAS_DBUS_IFACE_INTERFACE,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                     "ApIsolate",
                                     g_variant_new_string(ap_isolation ? "1" : "0"),
                                     DBUS_TIMEOUT_MSEC,
@@ -2654,7 +2706,7 @@ scan_request_cb(GObject *source, GAsyncResult *result, gpointer user_data)
         _LOGD("request-scan: request cancelled");
     else {
         if (error) {
-            if (nm_dbus_error_is(error, "fi.w1.wpa_supplicant1.Interface.ScanError"))
+            if (nm_dbus_error_is(error, "fi.w1.wpa_supplicant1.Interface.ScanError", "fi.w1.wpa_supplicant1_s1g.Interface.ScanError"))
                 _LOGD("request-scan: could not get scan request result: %s", error->message);
             else {
                 g_dbus_error_strip_remote_error(error);
@@ -2725,9 +2777,10 @@ nm_supplicant_interface_request_scan(NMSupplicantInterface                   *se
         nm_assert(!cancellable);
         cancellable = priv->main_cancellable;
     }
-
+  
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call(self,
-                          NM_WPAS_DBUS_IFACE_INTERFACE,
+                          is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                           "Scan",
                           g_variant_new("(a{sv})", &builder),
                           G_VARIANT_TYPE("()"),
@@ -2804,8 +2857,10 @@ nm_supplicant_interface_p2p_start_find(NMSupplicantInterface *self, guint timeou
     g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
     g_variant_builder_add(&builder, "{sv}", "Timeout", g_variant_new_int32(timeout));
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple(self,
-                                 NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                 is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                  "Find",
                                  g_variant_new("(a{sv})", &builder),
                                  G_VARIANT_TYPE("()"),
@@ -2817,8 +2872,10 @@ nm_supplicant_interface_p2p_stop_find(NMSupplicantInterface *self)
 {
     g_return_if_fail(NM_IS_SUPPLICANT_INTERFACE(self));
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple(self,
-                                 NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                 is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                  "StopFind",
                                  NULL,
                                  G_VARIANT_TYPE("()"),
@@ -2847,8 +2904,10 @@ nm_supplicant_interface_p2p_connect(NMSupplicantInterface *self,
     g_variant_builder_add(&builder, "{sv}", "persistent", g_variant_new_boolean(FALSE));
     g_variant_builder_add(&builder, "{sv}", "go_intent", g_variant_new_int32(7));
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple(self,
-                                 NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                 is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                  "Connect",
                                  g_variant_new("(a{sv})", &builder),
                                  G_VARIANT_TYPE("(s)"),
@@ -2860,8 +2919,10 @@ nm_supplicant_interface_p2p_cancel_connect(NMSupplicantInterface *self)
 {
     g_return_if_fail(NM_IS_SUPPLICANT_INTERFACE(self));
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple_full(self,
-                                      NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                      is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                       "Cancel",
                                       NULL,
                                       G_VARIANT_TYPE("()"),
@@ -2874,8 +2935,10 @@ nm_supplicant_interface_p2p_disconnect(NMSupplicantInterface *self)
 {
     g_return_if_fail(NM_IS_SUPPLICANT_INTERFACE(self));
 
+    NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
     _dbus_connection_call_simple(self,
-                                 NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                 is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                  "Disconnect",
                                  NULL,
                                  G_VARIANT_TYPE("()"),
@@ -2898,7 +2961,7 @@ _properties_changed(NMSupplicantInterface *self,
     if (initial)
         priv->starting_pending_count--;
 
-    if ((initial || priv->is_ready_main) && nm_streq(interface_name, NM_WPAS_DBUS_IFACE_INTERFACE))
+    if ((initial || priv->is_ready_main) && (nm_streq(interface_name, NM_WPAS_DBUS_IFACE_INTERFACE) || nm_streq(interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_S1G)))
         is_main = TRUE;
     else if ((initial || priv->is_ready_p2p_device)
              && nm_streq(interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE)) {
@@ -3019,6 +3082,7 @@ _get_all_main_cb(GVariant *result, GError *error, gpointer user_data)
     if (result)
         g_variant_get(result, "(@a{sv})", &properties);
     _properties_changed(user_data, NM_WPAS_DBUS_IFACE_INTERFACE, properties, TRUE);
+    _properties_changed(user_data, NM_WPAS_DBUS_IFACE_INTERFACE_S1G, properties, TRUE);
 }
 
 static void
@@ -3032,6 +3096,7 @@ _get_all_p2p_device_cb(GVariant *result, GError *error, gpointer user_data)
     if (result)
         g_variant_get(result, "(@a{sv})", &properties);
     _properties_changed(user_data, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE, properties, TRUE);
+    _properties_changed(user_data, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G, properties, TRUE);
 }
 
 static void
@@ -3052,7 +3117,7 @@ _signal_handle(NMSupplicantInterface *self,
     NMSupplicantInterfacePrivate *priv = NM_SUPPLICANT_INTERFACE_GET_PRIVATE(self);
     const char                   *path;
 
-    if (nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE)) {
+    if (nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE) || nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_S1G) ) {
         if (!priv->is_ready_main)
             return;
 
@@ -3108,7 +3173,7 @@ _signal_handle(NMSupplicantInterface *self,
         return;
     }
 
-    if (nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE)) {
+    if (nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE) || nm_streq(signal_interface_name, NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G) ) {
         if (!priv->is_ready_p2p_device)
             return;
 
@@ -3341,6 +3406,10 @@ set_property(GObject *object, guint prop_id, const GValue *value, GParamSpec *ps
             nm_ref_string_ref(nm_supplicant_manager_get_dbus_name_owner(priv->supplicant_manager));
         nm_assert(NM_IS_REF_STRING(priv->name_owner));
 
+        priv->name_owner_s1g =
+            nm_ref_string_ref(nm_supplicant_manager_get_dbus_name_owner_s1g(priv->supplicant_manager));
+        nm_assert(NM_IS_REF_STRING(priv->name_owner_s1g));
+
         priv->global_capabilities =
             nm_supplicant_manager_get_global_capabilities(priv->supplicant_manager);
         break;
@@ -3409,10 +3478,12 @@ constructed(GObject *object)
     G_OBJECT_CLASS(nm_supplicant_interface_parent_class)->constructed(object);
 
     _LOGD("new supplicant interface %s on %s", priv->object_path->str, priv->name_owner->str);
+    bool is_s1g = is_path_s1g(priv->object_path->str);
+    const char* bus_name = is_s1g ? priv->name_owner_s1g->str : priv->name_owner->str;
 
     priv->properties_changed_id =
         nm_dbus_connection_signal_subscribe_properties_changed(priv->dbus_connection,
-                                                               priv->name_owner->str,
+                                                               bus_name,
                                                                priv->object_path->str,
                                                                NULL,
                                                                _properties_changed_cb,
@@ -3421,15 +3492,15 @@ constructed(GObject *object)
 
     priv->bss_properties_changed_id =
         nm_dbus_connection_signal_subscribe_properties_changed(priv->dbus_connection,
-                                                               priv->name_owner->str,
+                                                               bus_name,
                                                                NULL,
-                                                               NM_WPAS_DBUS_IFACE_BSS,
+                                                               is_s1g ? NM_WPAS_DBUS_IFACE_BSS_S1G : NM_WPAS_DBUS_IFACE_BSS,
                                                                _bss_properties_changed_cb,
                                                                self,
                                                                NULL);
 
     priv->signal_id = g_dbus_connection_signal_subscribe(priv->dbus_connection,
-                                                         priv->name_owner->str,
+                                                         bus_name,
                                                          NULL,
                                                          NULL,
                                                          priv->object_path->str,
@@ -3441,9 +3512,9 @@ constructed(GObject *object)
 
     /* Scan result aging parameters */
     nm_dbus_connection_call_set(priv->dbus_connection,
-                                priv->name_owner->str,
+                                bus_name,
                                 priv->object_path->str,
-                                NM_WPAS_DBUS_IFACE_INTERFACE,
+                                is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                 "BSSExpireAge",
                                 g_variant_new_uint32(250),
                                 DBUS_TIMEOUT_MSEC,
@@ -3451,9 +3522,9 @@ constructed(GObject *object)
                                 NULL,
                                 NULL);
     nm_dbus_connection_call_set(priv->dbus_connection,
-                                priv->name_owner->str,
+                                bus_name,
                                 priv->object_path->str,
-                                NM_WPAS_DBUS_IFACE_INTERFACE,
+                                is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                 "BSSExpireCount",
                                 g_variant_new_uint32(2),
                                 DBUS_TIMEOUT_MSEC,
@@ -3464,9 +3535,9 @@ constructed(GObject *object)
     if (_get_capability(priv, NM_SUPPL_CAP_TYPE_PMF) == NM_TERNARY_TRUE) {
         /* Initialize global PMF setting to 'optional' */
         nm_dbus_connection_call_set(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     priv->object_path->str,
-                                    NM_WPAS_DBUS_IFACE_INTERFACE,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                     "Pmf",
                                     g_variant_new_string("1"),
                                     DBUS_TIMEOUT_MSEC,
@@ -3496,9 +3567,9 @@ constructed(GObject *object)
 
     priv->starting_pending_count++;
     nm_dbus_connection_call_get_all(priv->dbus_connection,
-                                    priv->name_owner->str,
+                                    bus_name,
                                     priv->object_path->str,
-                                    NM_WPAS_DBUS_IFACE_INTERFACE,
+                                    is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE,
                                     5000,
                                     priv->main_cancellable,
                                     _get_all_main_cb,
@@ -3507,18 +3578,18 @@ constructed(GObject *object)
     if (_get_capability(priv, NM_SUPPL_CAP_TYPE_P2P) == NM_TERNARY_TRUE) {
         priv->peer_properties_changed_id =
             nm_dbus_connection_signal_subscribe_properties_changed(priv->dbus_connection,
-                                                                   priv->name_owner->str,
+                                                                   bus_name,
                                                                    NULL,
-                                                                   NM_WPAS_DBUS_IFACE_PEER,
+                                                                   is_s1g ? NM_WPAS_DBUS_IFACE_PEER_S1G : NM_WPAS_DBUS_IFACE_PEER,
                                                                    _peer_properties_changed_cb,
                                                                    self,
                                                                    NULL);
 
         priv->starting_pending_count++;
         nm_dbus_connection_call_get_all(priv->dbus_connection,
-                                        priv->name_owner->str,
+                                        bus_name,
                                         priv->object_path->str,
-                                        NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
+                                        is_s1g ? NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G : NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE,
                                         5000,
                                         priv->main_cancellable,
                                         _get_all_p2p_device_cb,
diff --git a/src/core/supplicant/nm-supplicant-manager.c b/src/core/supplicant/nm-supplicant-manager.c
index 3b80569..fe788c3 100644
--- a/src/core/supplicant/nm-supplicant-manager.c
+++ b/src/core/supplicant/nm-supplicant-manager.c
@@ -66,6 +66,12 @@ typedef struct {
     /* see nm_supplicant_manager_get_available(). */
     NMTernary available : 2;
 
+    /* S1G */
+    guint name_owner_changed_id_s1g;
+    GCancellable *get_name_owner_cancellable_s1g;
+    /* DBus name of the s1g supplicant */
+    NMRefString *name_owner_s1g; 
+
 } NMSupplicantManagerPrivate;
 
 struct _NMSupplicantManager {
@@ -144,6 +150,14 @@ nm_supplicant_manager_get_dbus_name_owner(NMSupplicantManager *self)
     return NM_SUPPLICANT_MANAGER_GET_PRIVATE(self)->name_owner;
 }
 
+NMRefString *
+nm_supplicant_manager_get_dbus_name_owner_s1g(NMSupplicantManager *self)
+{
+    g_return_val_if_fail(NM_IS_SUPPLICANT_MANAGER(self), NULL);
+    NMSupplicantManagerPrivate *priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
+    return NM_SUPPLICANT_MANAGER_GET_PRIVATE(self)->name_owner_s1g;
+}
+
 GDBusConnection *
 nm_supplicant_manager_get_dbus_connection(NMSupplicantManager *self)
 {
@@ -243,6 +257,7 @@ nm_supplicant_manager_set_wfd_ies(NMSupplicantManager *self, GBytes *wfd_ies)
     g_variant_builder_init(&params, G_VARIANT_TYPE("(ssv)"));
 
     g_variant_builder_add(&params, "s", NM_WPAS_DBUS_INTERFACE);
+    g_variant_builder_add(&params, "s", NM_WPAS_DBUS_INTERFACE_S1G);
     g_variant_builder_add(&params, "s", "WFDIEs");
     g_variant_builder_add_value(&params, g_variant_new_variant(nm_g_bytes_to_variant_ay(wfd_ies)));
 
@@ -258,6 +273,18 @@ nm_supplicant_manager_set_wfd_ies(NMSupplicantManager *self, GBytes *wfd_ies)
                            NULL,
                            on_supplicant_wfd_ies_set,
                            NULL);
+    g_dbus_connection_call(priv->dbus_connection,
+                           priv->name_owner_s1g->str,
+                           NM_WPAS_DBUS_PATH_S1G,
+                           DBUS_INTERFACE_PROPERTIES,
+                           "Set",
+                           g_variant_builder_end(&params),
+                           G_VARIANT_TYPE("()"),
+                           G_DBUS_CALL_FLAGS_NO_AUTO_START,
+                           3000,
+                           NULL,
+                           on_supplicant_wfd_ies_set,
+                           NULL);
 }
 
 /*****************************************************************************/
@@ -450,7 +477,7 @@ _create_iface_dbus_call_get_interface_cb(GObject *source, GAsyncResult *result,
         char ifname[NM_IFNAMSIZ];
 
         if (handle->create_iface_try_count < CREATE_IFACE_TRY_COUNT_MAX
-            && nm_dbus_error_is(error, NM_WPAS_ERROR_UNKNOWN_IFACE)
+            && nm_dbus_error_is(error, NM_WPAS_ERROR_UNKNOWN_IFACE, NM_WPAS_ERROR_UNKNOWN_IFACE_S1G)
             && nm_platform_if_indextoname(NM_PLATFORM_GET, handle->ifindex, ifname)) {
             /* Before, supplicant told us the interface existed. Was there a race?
              * Try again. */
@@ -477,6 +504,81 @@ _create_iface_dbus_call_get_interface_cb(GObject *source, GAsyncResult *result,
     _create_iface_add(self, handle, iface_path_str, FALSE);
 }
 
+static void
+_create_iface_dbus_call_create_interface_cb_s1g(GObject      *source,
+                                            GAsyncResult *result,
+                                            gpointer      user_data)
+{
+    GDBusConnection             *dbus_connection = G_DBUS_CONNECTION(source);
+    NMSupplMgrCreateIfaceHandle *handle          = user_data;
+    NMSupplicantManager         *self;
+    NMSupplicantManagerPrivate  *priv;
+    gs_unref_variant GVariant   *res   = NULL;
+    gs_free_error GError        *error = NULL;
+    const char                  *iface_path_str;
+    char                         ifname[NM_IFNAMSIZ];
+
+    res = g_dbus_connection_call_finish(dbus_connection, result, &error);
+
+    nm_shutdown_wait_obj_unregister(g_steal_pointer(&handle->shutdown_handle));
+
+    if (!res) {
+        if (handle->callback && ({
+                nm_assert(handle->self);
+                TRUE;
+            })
+            && nm_dbus_error_is(error, NM_WPAS_ERROR_EXISTS_ERROR_S1G)
+            && nm_platform_if_indextoname(NM_PLATFORM_GET, handle->ifindex, ifname)) {
+            self = handle->self;
+            _LOGT("create-iface[" NM_HASH_OBFUSCATE_PTR_FMT
+                  "]: D-Bus call failed to create interface. Try to get existing interface (ifname "
+                  "\"%s\")",
+                  NM_HASH_OBFUSCATE_PTR(handle),
+                  ifname);
+            _create_iface_dbus_call_get_interface(self, handle, ifname);
+            return;
+        }
+        g_clear_object(&handle->cancellable);
+        _LOGT("create-iface[" NM_HASH_OBFUSCATE_PTR_FMT "]: D-Bus call failed: %s",
+              NM_HASH_OBFUSCATE_PTR(handle),
+              error->message);
+        _create_iface_complete(handle, NULL, error);
+        return;
+    }
+
+    g_clear_object(&handle->cancellable);
+
+    self = handle->self;
+    priv = self ? NM_SUPPLICANT_MANAGER_GET_PRIVATE(self) : NULL;
+
+    g_variant_get(res, "(&o)", &iface_path_str);
+
+    if (!handle->callback || priv->name_owner_s1g != handle->name_owner) {
+        if (!handle->callback) {
+            _LOGT("create-iface[" NM_HASH_OBFUSCATE_PTR_FMT
+                  "]: request already cancelled but still remove interface %s in %s",
+                  NM_HASH_OBFUSCATE_PTR(handle),
+                  iface_path_str,
+                  priv->name_owner_s1g ->str);
+            nm_utils_error_set(&error, NM_UTILS_ERROR_UNKNOWN, "Request already cancelled");
+        } else {
+            _LOGT("create-iface[" NM_HASH_OBFUSCATE_PTR_FMT
+                  "]: name owner changed, still remove interface %s in %s",
+                  NM_HASH_OBFUSCATE_PTR(handle),
+                  iface_path_str,
+                  priv->name_owner_s1g ->str);
+            nm_utils_error_set(&error,
+                               NM_UTILS_ERROR_UNKNOWN,
+                               "The name owner changed since creating the interface");
+        }
+        _dbus_call_remove_interface(dbus_connection, priv->name_owner_s1g->str, iface_path_str);
+        _create_iface_complete(handle, NULL, error);
+        return;
+    }
+
+    _create_iface_add(self, handle, iface_path_str, TRUE);
+}
+
 static void
 _create_iface_dbus_call_create_interface_cb(GObject      *source,
                                             GAsyncResult *result,
@@ -552,6 +654,25 @@ _create_iface_dbus_call_create_interface_cb(GObject      *source,
     _create_iface_add(self, handle, iface_path_str, TRUE);
 }
 
+static bool is_iface_mm(const char *ifname){
+      char path[PATH_MAX];
+      char buf[PATH_MAX];
+      sprintf(path, "/sys/class/net/%s/device/driver", ifname);
+      ssize_t len = readlink(path, buf, sizeof(buf) - 1);
+      if (len == -1) {
+          return false;
+      }
+      buf[len] = '\0';  // Null-terminate the buffer
+      const char* MM = "morse";
+      for(int i = 0; i < len; i++){
+        int cmp = strncmp(buf+i, MM, strlen(MM));
+        if (cmp == 0){
+          return true;
+        }
+      }
+      return false;
+}
+
 static void
 _create_iface_dbus_call_get_interface(NMSupplicantManager         *self,
                                       NMSupplMgrCreateIfaceHandle *handle,
@@ -561,19 +682,36 @@ _create_iface_dbus_call_get_interface(NMSupplicantManager         *self,
 
     nm_assert(handle->cancellable);
     nm_assert(!handle->shutdown_handle);
-
-    g_dbus_connection_call(priv->dbus_connection,
-                           priv->name_owner->str,
-                           NM_WPAS_DBUS_PATH,
-                           NM_WPAS_DBUS_INTERFACE,
-                           "GetInterface",
-                           g_variant_new("(s)", ifname),
-                           G_VARIANT_TYPE("(o)"),
-                           G_DBUS_CALL_FLAGS_NONE,
-                           5000,
-                           handle->cancellable,
-                           _create_iface_dbus_call_get_interface_cb,
-                           handle);
+    
+    const bool is_mm = is_iface_mm(ifname);
+    if(is_mm){
+      g_dbus_connection_call(priv->dbus_connection,
+                             priv->name_owner_s1g->str,
+                             NM_WPAS_DBUS_PATH_S1G,
+                             NM_WPAS_DBUS_INTERFACE_S1G,
+                             "GetInterface",
+                             g_variant_new("(s)", ifname),
+                             G_VARIANT_TYPE("(o)"),
+                             G_DBUS_CALL_FLAGS_NONE,
+                             5000,
+                             handle->cancellable,
+                             _create_iface_dbus_call_get_interface_cb,
+                             handle);
+    }
+    else{
+      g_dbus_connection_call(priv->dbus_connection,
+                             priv->name_owner->str,
+                             NM_WPAS_DBUS_PATH,
+                             NM_WPAS_DBUS_INTERFACE,
+                             "GetInterface",
+                             g_variant_new("(s)", ifname),
+                             G_VARIANT_TYPE("(o)"),
+                             G_DBUS_CALL_FLAGS_NONE,
+                             5000,
+                             handle->cancellable,
+                             _create_iface_dbus_call_get_interface_cb,
+                             handle);
+    }
 }
 
 static void
@@ -584,35 +722,54 @@ _create_iface_dbus_call_create_interface(NMSupplicantManager         *self,
     NMSupplicantManagerPrivate *priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
     GVariantBuilder             builder;
 
-    nm_assert(priv->name_owner == handle->name_owner);
+    nm_assert((priv->name_owner == handle->name_owner) || (priv->name_owner_s1g == handle->name_owner));
     nm_assert(handle->cancellable);
     nm_assert(!handle->shutdown_handle);
     nm_assert(handle->create_iface_try_count <= CREATE_IFACE_TRY_COUNT_MAX);
 
-    g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
-    g_variant_builder_add(&builder,
-                          "{sv}",
-                          "Driver",
-                          g_variant_new_string(nm_supplicant_driver_to_string(handle->driver)));
-    g_variant_builder_add(&builder, "{sv}", "Ifname", g_variant_new_string(ifname));
-
-    handle->shutdown_handle = nm_shutdown_wait_obj_register_cancellable_full(
-        handle->cancellable,
-        g_strdup_printf("wpas-create-" NM_HASH_OBFUSCATE_PTR_FMT, NM_HASH_OBFUSCATE_PTR(handle)),
-        TRUE);
-    handle->create_iface_try_count++;
-    g_dbus_connection_call(priv->dbus_connection,
-                           handle->name_owner->str,
-                           NM_WPAS_DBUS_PATH,
-                           NM_WPAS_DBUS_INTERFACE,
-                           "CreateInterface",
-                           g_variant_new("(a{sv})", &builder),
-                           G_VARIANT_TYPE("(o)"),
-                           G_DBUS_CALL_FLAGS_NONE,
-                           5000,
-                           handle->cancellable,
-                           _create_iface_dbus_call_create_interface_cb,
-                           handle);
+      g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
+      g_variant_builder_add(&builder,
+                            "{sv}",
+                            "Driver",
+                            g_variant_new_string(nm_supplicant_driver_to_string(handle->driver)));
+      g_variant_builder_add(&builder, "{sv}", "Ifname", g_variant_new_string(ifname));
+
+      handle->shutdown_handle = nm_shutdown_wait_obj_register_cancellable_full(
+          handle->cancellable,
+          g_strdup_printf("wpas-create-" NM_HASH_OBFUSCATE_PTR_FMT, NM_HASH_OBFUSCATE_PTR(handle)),
+          TRUE);
+      handle->create_iface_try_count++;
+    const bool is_mm = is_iface_mm(ifname);
+    if(is_mm){
+      g_dbus_connection_call(priv->dbus_connection,
+                             priv->name_owner_s1g->str,
+                             NM_WPAS_DBUS_PATH_S1G,
+                             NM_WPAS_DBUS_INTERFACE_S1G,
+                             "CreateInterface",
+                             g_variant_new("(a{sv})", &builder),
+                             G_VARIANT_TYPE("(o)"),
+                             G_DBUS_CALL_FLAGS_NONE,
+                             5000,
+                             handle->cancellable,
+                             _create_iface_dbus_call_create_interface_cb_s1g,
+                             handle);
+    }
+    else
+    {
+      g_dbus_connection_call(priv->dbus_connection,
+                             handle->name_owner->str,
+                             NM_WPAS_DBUS_PATH,
+                             NM_WPAS_DBUS_INTERFACE,
+                             "CreateInterface",
+                             g_variant_new("(a{sv})", &builder),
+                             G_VARIANT_TYPE("(o)"),
+                             G_DBUS_CALL_FLAGS_NONE,
+                             5000,
+                             handle->cancellable,
+                             _create_iface_dbus_call_create_interface_cb,
+                             handle);
+    }
+
 }
 
 static void
@@ -1099,6 +1256,123 @@ _available_reset_cb(gpointer user_data)
 
 /*****************************************************************************/
 
+static void
+name_owner_changed_s1g(NMSupplicantManager *self, const char *name_owner, gboolean first_time)
+{
+    NMSupplicantManagerPrivate *priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
+    NMTernary                   available;
+    gboolean                    available_changed = FALSE;
+
+    nm_assert(!priv->get_name_owner_cancellable_s1g);
+    nm_assert(!name_owner || name_owner[0]);
+    nm_assert((first_time && !priv->name_owner_s1g)
+              || (!first_time && (!!priv->name_owner_s1g) != (!!name_owner)));
+
+    if (first_time) {
+        _LOGD("wpa_supplicant name owner %s%s%s (%srunning)",
+              NM_PRINT_FMT_QUOTE_STRING(name_owner),
+              name_owner ? "" : "not ");
+    } else {
+        _LOGD("wpa_supplicant name owner \"%s\" %s (%srunning)",
+              name_owner ?: priv->name_owner->str,
+              name_owner ? "disappeared" : "appeared",
+              name_owner ? "" : "not ");
+    }
+
+    nm_ref_string_unref(priv->name_owner_s1g);
+    priv->name_owner_s1g = nm_ref_string_new(name_owner);
+
+    nm_clear_g_dbus_connection_signal(priv->dbus_connection, &priv->interface_removed_id);
+
+    if (name_owner) {
+        if (nm_clear_g_source(&priv->poke_name_owner_timeout_id))
+            _LOGT("poke service \"%s\" completed with name owner change", NM_WPAS_DBUS_SERVICE_S1G);
+        nm_clear_g_cancellable(&priv->poke_name_owner_cancellable);
+    }
+
+    nm_clear_g_cancellable(&priv->get_capabilities_cancellable);
+
+    priv->capabilities = NM_SUPPL_CAP_MASK_NONE;
+    if (priv->name_owner_s1g) {
+        priv->get_capabilities_cancellable = g_cancellable_new();
+        nm_dbus_connection_call_get_all(priv->dbus_connection,
+                                        priv->name_owner_s1g->str,
+                                        NM_WPAS_DBUS_PATH_S1G,
+                                        NM_WPAS_DBUS_INTERFACE_S1G,
+                                        5000,
+                                        priv->get_capabilities_cancellable,
+                                        _dbus_get_capabilities_cb,
+                                        self);
+        priv->interface_removed_id = g_dbus_connection_signal_subscribe(priv->dbus_connection,
+                                                                        priv->name_owner_s1g->str,
+                                                                        NM_WPAS_DBUS_INTERFACE_S1G,
+                                                                        "InterfaceRemoved",
+                                                                        NULL,
+                                                                        NULL,
+                                                                        G_DBUS_SIGNAL_FLAGS_NONE,
+                                                                        _dbus_interface_removed_cb,
+                                                                        self,
+                                                                        NULL);
+    }
+
+    /* if supplicant is running (has a name owner), we may use it.
+     * If this is the first time, and supplicant is not running, we
+     * may also use it (and assume that we probably could D-Bus activate
+     * it).
+     *
+     * Otherwise, somebody else stopped supplicant. It's no longer useable to
+     * us and we block auto starting it. The user has to start the service...
+     *
+     * Actually, below we reset the hard block after a short timeout. This
+     * causes the caller to notify that supplicant may now by around and
+     * retry to D-Bus activate it. */
+    if (priv->name_owner_s1g)
+        available = NM_TERNARY_TRUE;
+    else if (first_time)
+        available = NM_TERNARY_DEFAULT;
+    else
+        available = NM_TERNARY_FALSE;
+
+    if (priv->available != available) {
+        priv->available = available;
+        _LOGD("supplicant is now %savailable",
+              available == FALSE ? "not " : (available == TRUE ? "" : "maybe "));
+        available_changed = TRUE;
+
+        nm_clear_g_source(&priv->available_reset_id);
+        if (available == NM_TERNARY_FALSE) {
+            /* reset the availability from a hard "no" to a "maybe" in a bit. */
+            priv->available_reset_id = g_timeout_add_seconds(60, _available_reset_cb, self);
+        }
+    }
+
+    _supp_iface_remove_all(self, TRUE, "name-owner changed");
+
+    if (!priv->name_owner) {
+        if (priv->poke_name_owner_timeout_id) {
+            /* we are still poking for the service to start. Don't cancel
+             * the pending create requests just yet. */
+        } else {
+            gs_free_error GError *local_error = NULL;
+
+            /* When we loose the name owner, we fail all pending creation requests. */
+            nm_utils_error_set(&local_error, NM_UTILS_ERROR_UNKNOWN, "Name owner lost");
+            _create_iface_proceed_all(self, local_error);
+        }
+    } else {
+        /* We got a name-owner, but we don't do anything. Instead let
+         * _dbus_get_capabilities_cb() complete and kick of the create-iface
+         * handles.
+         *
+         * Note that before the first name-owner change, all create-iface
+         * requests fail right away. So we don't have to handle them here
+         * (by starting to poke the service). */
+    }
+
+    if (available_changed)
+        g_signal_emit(self, signals[AVAILABLE_CHANGED], 0);
+}
+
 static void
 name_owner_changed(NMSupplicantManager *self, const char *name_owner, gboolean first_time)
 {
@@ -1216,6 +1490,40 @@ name_owner_changed(NMSupplicantManager *self, const char *name_owner, gboolean f
         g_signal_emit(self, signals[AVAILABLE_CHANGED], 0);
 }
 
+static void
+name_owner_changed_cb_s1g(GDBusConnection *connection,
+                      const char      *sender_name,
+                      const char      *object_path,
+                      const char      *interface_name,
+                      const char      *signal_name,
+                      GVariant        *parameters,
+                      gpointer         user_data)
+{
+    gs_unref_object NMSupplicantManager *self = g_object_ref(user_data);
+    NMSupplicantManagerPrivate          *priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
+    const char                          *name_owner;
+
+    if (!g_variant_is_of_type(parameters, G_VARIANT_TYPE("(sss)")))
+        return;
+
+    if (priv->get_name_owner_cancellable_s1g)
+        return;
+
+    g_variant_get(parameters, "(&s&s&s)", NULL, NULL, &name_owner);
+
+    name_owner = nm_str_not_empty(name_owner);
+
+    if (nm_streq0(name_owner, nm_ref_string_get_str(priv->name_owner_s1g)))
+        return;
+
+    if (name_owner && priv->name_owner_s1g) {
+        /* odd, we directly switch from one name owner to the next. Can't allow that.
+         * First clear the name owner before resetting. */
+        name_owner_changed(self, NULL, FALSE);
+    }
+    name_owner_changed_s1g(user_data, name_owner, FALSE);
+}
+
 static void
 name_owner_changed_cb(GDBusConnection *connection,
                       const char      *sender_name,
@@ -1250,6 +1558,23 @@ name_owner_changed_cb(GDBusConnection *connection,
     name_owner_changed(user_data, name_owner, FALSE);
 }
 
+static void
+get_name_owner_cb_s1g(const char *name_owner, GError *error, gpointer user_data)
+{
+    NMSupplicantManager        *self = user_data;
+    NMSupplicantManagerPrivate *priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
+
+    if (!name_owner && nm_utils_error_is_cancelled(error))
+        return;
+
+    self = user_data;
+    priv = NM_SUPPLICANT_MANAGER_GET_PRIVATE(self);
+
+    g_clear_object(&priv->get_name_owner_cancellable_s1g);
+
+    name_owner_changed_s1g(self, nm_str_not_empty(name_owner), TRUE);
+}
+
 static void
 get_name_owner_cb(const char *name_owner, GError *error, gpointer user_data)
 {
@@ -1301,6 +1626,21 @@ nm_supplicant_manager_init(NMSupplicantManager *self)
                                            priv->get_name_owner_cancellable,
                                            get_name_owner_cb,
                                            self);
+
+    /* S1G */
+    priv->name_owner_changed_id_s1g =
+        nm_dbus_connection_signal_subscribe_name_owner_changed(priv->dbus_connection,
+                                                               NM_WPAS_DBUS_SERVICE_S1G,
+                                                               name_owner_changed_cb_s1g,
+                                                               self,
+                                                               NULL);
+    priv->get_name_owner_cancellable_s1g = g_cancellable_new();
+    nm_dbus_connection_call_get_name_owner(priv->dbus_connection,
+                                           NM_WPAS_DBUS_SERVICE_S1G,
+                                           -1,
+                                           priv->get_name_owner_cancellable_s1g,
+                                           get_name_owner_cb_s1g,
+                                           self);
 }
 
 static void
@@ -1324,9 +1664,15 @@ dispose(GObject *object)
     nm_clear_g_dbus_connection_signal(priv->dbus_connection, &priv->interface_removed_id);
     nm_clear_g_dbus_connection_signal(priv->dbus_connection, &priv->name_owner_changed_id);
 
+    /* S1G */
+    nm_clear_g_dbus_connection_signal(priv->dbus_connection, &priv->name_owner_changed_id_s1g);
+
     nm_clear_g_cancellable(&priv->get_name_owner_cancellable);
     nm_clear_g_cancellable(&priv->get_capabilities_cancellable);
 
+    /* S1G */
+    nm_clear_g_cancellable(&priv->get_name_owner_cancellable_s1g);
+
     G_OBJECT_CLASS(nm_supplicant_manager_parent_class)->dispose(object);
 
     g_clear_object(&priv->dbus_connection);
diff --git a/src/core/supplicant/nm-supplicant-manager.h b/src/core/supplicant/nm-supplicant-manager.h
index b3284b7..45dab1c 100644
--- a/src/core/supplicant/nm-supplicant-manager.h
+++ b/src/core/supplicant/nm-supplicant-manager.h
@@ -34,6 +34,7 @@ NMTernary nm_supplicant_manager_is_available(NMSupplicantManager *self);
 
 GDBusConnection *nm_supplicant_manager_get_dbus_connection(NMSupplicantManager *self);
 NMRefString     *nm_supplicant_manager_get_dbus_name_owner(NMSupplicantManager *self);
+NMRefString     *nm_supplicant_manager_get_dbus_name_owner_s1g(NMSupplicantManager *self);
 NMSupplCapMask   nm_supplicant_manager_get_global_capabilities(NMSupplicantManager *self);
 
 void nm_supplicant_manager_set_wfd_ies(NMSupplicantManager *self, GBytes *wfd_ies);
@@ -46,6 +47,8 @@ typedef void (*NMSupplicantManagerCreateInterfaceCb)(NMSupplicantManager
                                                      GError                      *error,
                                                      gpointer                     user_data);
 
+bool is_s1g(NMSupplicantManager *supplicant_manager);
+
 NMSupplMgrCreateIfaceHandle *
 nm_supplicant_manager_create_interface(NMSupplicantManager                 *self,
                                        int                                  ifindex,
diff --git a/src/core/supplicant/nm-supplicant-types.h b/src/core/supplicant/nm-supplicant-types.h
index 870a53c..94844f3 100644
--- a/src/core/supplicant/nm-supplicant-types.h
+++ b/src/core/supplicant/nm-supplicant-types.h
@@ -9,9 +9,13 @@
 #include "c-list/src/c-list.h"
 #include "libnm-base/nm-base.h"
 
-#define NM_WPAS_DBUS_SERVICE   "fi.w1.wpa_supplicant1"
+#define NM_WPAS_DBUS_SERVICE    "fi.w1.wpa_supplicant1"
 #define NM_WPAS_DBUS_PATH      "/fi/w1/wpa_supplicant1"
-#define NM_WPAS_DBUS_INTERFACE "fi.w1.wpa_supplicant1"
+#define NM_WPAS_DBUS_INTERFACE  "fi.w1.wpa_supplicant1"
+
+#define NM_WPAS_DBUS_SERVICE_S1G   "fi.w1.wpa_supplicant1_s1g"
+#define NM_WPAS_DBUS_PATH_S1G     "/fi/w1/wpa_supplicant1_s1g"
+#define NM_WPAS_DBUS_INTERFACE_S1G "fi.w1.wpa_supplicant1_s1g"
 
 #if HAVE_WEXT
 #define NM_WPAS_DEFAULT_WIFI_DRIVER "nl80211,wext"
@@ -30,6 +34,17 @@
 #define NM_WPAS_ERROR_EXISTS_ERROR              NM_WPAS_DBUS_INTERFACE ".InterfaceExists"
 #define NM_WPAS_ERROR_UNKNOWN_IFACE             NM_WPAS_DBUS_INTERFACE ".InterfaceUnknown"
 
+#define NM_WPAS_DBUS_IFACE_INTERFACE_S1G            NM_WPAS_DBUS_INTERFACE_S1G ".Interface"
+#define NM_WPAS_DBUS_IFACE_INTERFACE_WPS_S1G        NM_WPAS_DBUS_INTERFACE_S1G ".Interface.WPS"
+#define NM_WPAS_DBUS_IFACE_INTERFACE_P2P_DEVICE_S1G NM_WPAS_DBUS_INTERFACE_S1G ".Interface.P2PDevice"
+#define NM_WPAS_DBUS_IFACE_BSS_S1G                  NM_WPAS_DBUS_INTERFACE_S1G ".BSS"
+#define NM_WPAS_DBUS_IFACE_PEER_S1G                 NM_WPAS_DBUS_INTERFACE_S1G ".Peer"
+#define NM_WPAS_DBUS_IFACE_GROUP_S1G                NM_WPAS_DBUS_INTERFACE_S1G ".Group"
+#define NM_WPAS_DBUS_IFACE_NETWORK_S1G              NM_WPAS_DBUS_INTERFACE_S1G ".Network"
+#define NM_WPAS_ERROR_INVALID_IFACE_S1G             NM_WPAS_DBUS_INTERFACE_S1G ".InvalidInterface"
+#define NM_WPAS_ERROR_EXISTS_ERROR_S1G              NM_WPAS_DBUS_INTERFACE_S1G ".InterfaceExists"
+#define NM_WPAS_ERROR_UNKNOWN_IFACE_S1G             NM_WPAS_DBUS_INTERFACE_S1G ".InterfaceUnknown"
+
 typedef struct _NMSupplicantManager   NMSupplicantManager;
 typedef struct _NMSupplicantInterface NMSupplicantInterface;
 typedef struct _NMSupplicantConfig    NMSupplicantConfig;
